### PR TITLE
Improve solution of vector problems

### PR DESCRIPTION
**Pull request by yguclu**
Improve `Equation.solve()`:

- In addition to `cg` (conjugate gradient = CG) solver, allow `pcg` (pre-conditioned CG) and `bicg` (bi-CG);
- Fix bug when 'info = True' argument is passed to function;
- Correctly apply non-homogeneous Dirichlet boundary conditions to vector quantities.

Further improvements:

- Fix bug in `BlockMatrix.tosparse()` when `scipy.sparse.bmat()` computes wrong shape because of too many empty blocks;
- Extend 2D Stokes unit tests (add non-homogeneous test, verify results with Scipy).

### CODE REVIEW

File: psydac/api/tests/test_api_2d_compatible_spaces.py [hunk 1]
```diff
@@ -86,95 +91,126 @@ def run_system_1_2d_dir(f0, sol, ncells, degree):
...
+    a  = BilinearForm(((u, p), (v, q)), int_0(inner(grad(u), grad(v)) - div(u)*q - p*div(v)) )
+    l  = LinearForm((v, q), int_0(dot(f, v)))
 
-    error = Matrix([F[0]-u1, F[1]-u2])
-    l2norm_F = Norm(error, domain, kind='l2')
+    if homogeneous:
+        bc = EssentialBC(u, 0, domain.boundary)
+    else:
+        bc = EssentialBC(u, ue, domain.boundary)
+
```
Comment (campospinto) [inline psydac/api/tests/test_api_2d_compatible_spaces.py:114]: for documentation purposes we could add a remark specifying that the argument could be a function only defined on the boundary (instead of ue which is not known in a usual problem)

File: psydac/api/tests/test_api_2d_compatible_spaces.py [hunk 2]
```diff
@@ -86,95 +91,126 @@ def run_system_1_2d_dir(f0, sol, ncells, degree):
...
+    print()
+    print('Relative l2_error(u) = {}'.format(l2_error_u / l2_norm_ue))
+    print('Relative l2_error(p) = {}'.format(l2_error_p / l2_norm_pe))
+    print('Average(p) = {}'.format(p_avg))
+
+    return locals()
 
 #==============================================================================
-def run_system_3_2d_dir(uex, f, alpha, ncells, degree):
+def run_maxwell_time_harmonic_2d_dir(uex, f, alpha, ncells, degree):
```
Comment (campospinto) [inline psydac/api/tests/test_api_2d_compatible_spaces.py:214]: same comment as above (cite reference if relevant), and also for other equations below

File: psydac/api/tests/test_api_2d_compatible_spaces.py [hunk 3]
```diff
@@ -235,34 +271,102 @@ def run_system_3_2d_dir(uex, f, alpha, ncells, degree):
...
+
+    kwargs = dict(dim=2, logical=True)
+    a = TerminalExpr(-laplace(ue), **kwargs)
+    b = TerminalExpr(    grad(pe), **kwargs)
+    c = TerminalExpr(   Matrix(f), **kwargs)
+    err = (a.T + b - c).simplify()
+
+    assert err[0] == 0
+    assert err[1] == 0
+    # ...
```
Comment (campospinto) [inline psydac/api/tests/test_api_2d_compatible_spaces.py:317]: this is nice !

**Review by campospinto — COMMENTED** [review]
Very nice improvements, thanks! I added a few questions.

File: psydac/api/tests/test_api_2d_compatible_spaces.py [hunk 4]
```diff
@@ -1,28 +1,33 @@
...
 from psydac.api.discretization import discretize
 from psydac.linalg.utilities   import array_to_stencil
-
-from psydac.linalg.iterative_solvers import pcg
+from psydac.linalg.iterative_solvers import pcg, bicg
 
 #==============================================================================
-def run_system_1_2d_dir(f0, sol, ncells, degree):
+def run_poisson_mixed_form_2d_dir(f0, sol, ncells, degree):
     # ... abstract model
```
[the remark this answers is not in the document]
Comment (yguclu) [reply psydac/api/tests/test_api_2d_compatible_spaces.py:31]: I am not aware of a reference for this manufacture solution. Maybe @saidctb knows?

[the remark this answers is not in the document]
Comment (yguclu) [reply psydac/api/tests/test_api_2d_compatible_spaces.py:31]: By the way, this function get the manufactured solution as an input. The references should be cited in the `test_` functions.

[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/api/tests/test_api_2d_compatible_spaces.py:31]: no, the manufactured solution is the same for the poison problem.

File: psydac/api/tests/test_api_2d_compatible_spaces.py [hunk 5]
```diff
@@ -86,95 +91,126 @@ def run_system_1_2d_dir(f0, sol, ncells, degree):
     return l2_error
 
 #==============================================================================
-def run_system_2_2d_dir(f1, f2,u1, u2, ncells, degree):
-    # ... abstract model
-    domain = Square()
+def run_stokes_2d_dir(domain, f, ue, pe, *, homogeneous, ncells, degree, scipy=False):
```
[the remark this answers is not in the document]
Comment (yguclu) [reply psydac/api/tests/test_api_2d_compatible_spaces.py:94]: Maybe @saidctb knows a reference for the first Stokes problem? I created the second one by modifying what he's using in his thesis draft (which he created).

File: psydac/api/tests/test_api_2d_compatible_spaces.py [hunk 6]
```diff
@@ -86,95 +91,126 @@ def run_system_1_2d_dir(f0, sol, ncells, degree):
...
     int_0 = lambda expr: integral(domain , expr)
 
-    
-    a  = BilinearForm(((u,p),(v,q)), int_0(inner(grad(u),grad(v)) + div(u)*q - p*div(v)) )
-    l  = LinearForm((v,q), int_0(f1*v[0]+f2*v[1]+q))
-    
-    bc = EssentialBC(u, 0, domain.boundary)
-    equation = find([u,p], forall=[v,q], lhs=a((u,p),(v,q)), rhs=l(v,q), bc=bc)
+    a  = BilinearForm(((u, p), (v, q)), int_0(inner(grad(u), grad(v)) - div(u)*q - p*div(v)) )
+    l  = LinearForm((v, q), int_0(dot(f, v)))
```
[the remark this answers is not in the document]
Comment (yguclu) [reply psydac/api/tests/test_api_2d_compatible_spaces.py:108]: This should be right, as the strong form of the Stokes equations only has a forcing term on the momentum equation. I need to double check.

**Comment by saidctb** [comment]
ca we add scipy into the  `solve` method and do `array_to_stencil` internally

File: psydac/api/tests/test_api_2d_compatible_spaces.py [hunk 7]
```diff
@@ -86,95 +91,126 @@ def run_system_1_2d_dir(f0, sol, ncells, degree):
...
+    equation = find((u, p), forall=(v, q), lhs=a((u, p), (v, q)), rhs=l(v, q), bc=bc)
 
     # ... create the computational domain from a topological domain
     domain_h = discretize(domain, ncells=ncells)
+
     # ... discrete spaces
     V1h = discretize(V1, domain_h, degree=degree)
-    V2h = discretize(V2, domain_h, degree=degree)
+    V2h = discretize(V2, domain_h, degree=degree, basis='M')
+    Xh  = discretize(X , domain_h, degree=degree, basis='M')
```
[the remark this answers is not in the document]
Comment (yguclu) [reply psydac/api/tests/test_api_2d_compatible_spaces.py:123]: The basis argument 'M' only causes a renormalization of the basis functions for the 1D bases where the degree is reduced. 

The polynomial degree is reduced because we declared the space `V2` to be of kind `l2`.

There are lots of assumptions in the underlying construction, which should be made explicit and cleaned up...

File: psydac/api/tests/test_api_2d_compatible_spaces.py [hunk 8]
```diff
@@ -86,95 +91,126 @@ def run_system_1_2d_dir(f0, sol, ncells, degree):
...
-    M[0,0][e11,:,0,0] = 1.
-    M[0,0][:,e12,0,0] = 1.
-    M[1,1][0,:,0,0] = 1.
-    M[1,1][:,0,0,0] = 1.
-    M[1,1][e21,:,0,0] = 1.
-    M[1,1][:,e22,0,0] = 1.
-    M = M.toarray()
-    
-    rhs   = ah.linear_system.rhs.toarray()
+        if not homogeneous:
```
[the remark this answers is not in the document]
Comment (yguclu) [reply psydac/api/tests/test_api_2d_compatible_spaces.py:137]: If the user wants to use Scipy, he/she will have to handle the boundary conditions manually as shown here.

File: psydac/api/tests/test_api_2d_compatible_spaces.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (campospinto) [reply psydac/api/tests/test_api_2d_compatible_spaces.py:123]: I understand. Indeed by looking at how V1 and V2 are defined and used, it seems that `kind` indicate the smoothness that is required on the functions. So yes, the assumptions in the construction should be clarified. 

This should probably be handled in a separate PR. I will update the issue #101 to include the construction of single spaces and the role of the `kind` parameter

File: psydac/api/tests/test_api_2d_compatible_spaces.py (hunk 2 above)
Comment (yguclu) [reply psydac/api/tests/test_api_2d_compatible_spaces.py:214]: @saidctb Is there a reference for this test case?

**Comment by campospinto** [comment]
> ca we add scipy into the `solve` method and do `array_to_stencil` internally
I am not sure this is a good idea. One interest of the numpy/scipy conversion is to allow a user to rely on general scipy libraries, which greatly improves the prototyping capabilities of Psydac. These conversions should be visible in these tests/examples, not hidden inside the psydac solvers.

**Comment by campospinto** [comment]
I think the only remaining point was to quote references for the tested problems, when available. Since this does not seem to be the case, the PR is ok for me.

File: psydac/api/tests/test_api_2d_compatible_spaces.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (yguclu) [reply psydac/api/tests/test_api_2d_compatible_spaces.py:108]: I confirm that the formulas in the code look correct to me. Accordingly, I can only conclude that the previous version contained a bug. Or maybe this was an attempt at stabilizing the solution @saidctb?

File: psydac/api/tests/test_api_2d_compatible_spaces.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (yguclu) [reply psydac/api/tests/test_api_2d_compatible_spaces.py:137]: @campospinto If the user does not want to use Scipy, then the flag is not read again. Do you have a different suggestion?

File: psydac/api/tests/test_api_2d_compatible_spaces.py (hunk 1 above)
Comment (yguclu) [reply psydac/api/tests/test_api_2d_compatible_spaces.py:114]: Indeed the argument is a symbolic expression `g(x, y)`, which will only be evaluated at the boundary. Could this expression be valid only at the boundary? It is hard to see how such an expression could be constructed, given that it is symbolic.

@saidctb Do you remember if we accept a generic symbol for `ue`, which we can then substitute for a field?

File: psydac/api/tests/test_api_2d_compatible_spaces.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/api/tests/test_api_2d_compatible_spaces.py:94]: no there is no reference

File: psydac/api/tests/test_api_2d_compatible_spaces.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/api/tests/test_api_2d_compatible_spaces.py:108]: yes, it was for stabilizing the solution

File: psydac/api/tests/test_api_2d_compatible_spaces.py (hunk 1 above)
Comment (saidctb) [reply psydac/api/tests/test_api_2d_compatible_spaces.py:114]: we never tested that, but it should be possible

File: psydac/api/tests/test_api_2d_compatible_spaces.py (hunk 2 above)
Comment (saidctb) [reply psydac/api/tests/test_api_2d_compatible_spaces.py:214]: no there is no reference

**Comment by saidctb** [comment]
> > ca we add scipy into the `solve` method and do `array_to_stencil` internally
> 
> I don't think this is a good idea. One interest of the numpy/scipy conversion is to allow a user to rely on general scipy libraries, which greatly improves the prototyping capabilities of Psydac. These conversions should be visible in these tests/examples, not hidden inside the psydac solvers.

we can have both, this is just a way to present it nicely

File: psydac/api/tests/test_api_2d_compatible_spaces.py (hunk 1 above)
Comment (campospinto) [reply psydac/api/tests/test_api_2d_compatible_spaces.py:114]: Sorry, my remark was not very clear. I meant: "specifying that the argument could be a function which only represents the boundary data". This boundary data can be defined outside of the boundary, but there it will not correspond to the exact solution.

File: psydac/api/tests/test_api_2d_compatible_spaces.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (campospinto) [reply psydac/api/tests/test_api_2d_compatible_spaces.py:137]: No, I think it's fine like that.

**Review by campospinto — APPROVED** [review]

File: psydac/api/tests/test_api_2d_compatible_spaces.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (yguclu) [reply psydac/api/tests/test_api_2d_compatible_spaces.py:108]: Interesting! Could you give me a reference so I can look into that?

File: psydac/api/tests/test_api_2d_compatible_spaces.py (hunk 1 above)
Comment (yguclu) [reply psydac/api/tests/test_api_2d_compatible_spaces.py:117]: I've added a comment with commit caf7b5f

File: psydac/api/tests/test_api_2d_compatible_spaces.py (hunk 3 above)
Comment (yguclu) [reply psydac/api/tests/test_api_2d_compatible_spaces.py:320]: :sweat_smile: Yes, it's nice to check that the exact solution is correct!

File: psydac/api/tests/test_api_2d_compatible_spaces.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/api/tests/test_api_2d_compatible_spaces.py:108]: I mean for the Lagrange multiplier

[the remark this answers is not in the document]
Comment (campospinto) [reply psydac/api/tests/test_api_2d_compatible_spaces.py:108]: For the readability of the code it is indeed better to point to a reference where the method is described. Otherwise (eg if the method is very simple), an explanation should be given in comments.

[the remark this answers is not in the document]
Comment (saidctb) [reply psydac/api/tests/test_api_2d_compatible_spaces.py:108]: it's deleted by the changes of Yaman

[the remark this answers is not in the document]
Comment (yguclu) [reply psydac/api/tests/test_api_2d_compatible_spaces.py:108]: @campospinto I think I understand now. At some point @saidctb had an additional term for ensuring that the average of the pressure was zero, through the use of a Lagrange multiplier. But this was handled completely by hand, and we have removed it here as these are unit tests for our API. Instead, we are just removing the average of the pressure field from the final solution.

**Review by saidctb — APPROVED** [review]
